src/ralph/lib/permissions/views.py

```diff
@@ -0,0 +1,53 @@
+# -*- coding: utf-8 -*-
+
+from django.contrib.auth import get_user_model
+from django.contrib.auth.models import Permission
+from django.contrib.contenttypes.models import ContentType
+from django.http import HttpResponseForbidden
+
+from ralph.admin.sites import ralph_site
+
+_permission_views = []
+
+
+class PermissionViewMetaClass(type):
+
+    def __new__(cls, name, bases, attrs):
+        codename = 'can_view_{}'.format(name.lower())
+
+        attrs['permision_codename'] = codename
+        new_class = super().__new__(cls, name, bases, attrs)
+        dispatch = getattr(new_class, 'dispatch', None)
+
+        def new_dispatch(self, request, *args, **kwargs):
+            model = kwargs.get('model', None)
+            if not model:
+                model = get_user_model()
+            perm_name = '{}.{}'.format(
+                model._meta.app_label, self.permision_codename
+            )
+            if not request.user.has_perm(perm_name):
+                return HttpResponseForbidden()
+            return dispatch(self, request, *args, **kwargs)
+
+        setattr(new_class, 'dispatch', new_dispatch)
+        _permission_views.append((new_class, codename))
+        return new_class
+
+
+def update_extra_view_permissions(**kwargs):
+    admin_classes = {}
+    for model, admin_class in ralph_site._registry.items():
+        for change_view in admin_class.change_views:
+            admin_classes[change_view] = model
+
+    for class_view, codename in _permission_views:
+        model = admin_classes.get(class_view, None)
+        if not model:
+            model = get_user_model()
+        ct = ContentType.objects.get_for_model(model)
+        Permission.objects.get_or_create(
```

Comment:
what with permissions that are no longer necessary? For example associated view is no longer existing in source code.